Qutrit Unitary

mudit2812 opened the pull request:
EDIT: **[See this comment for updated description of pull request](https://github.com/PennyLaneAI/pennylane/pull/2699#issuecomment-1159257360)**
EDIT 2: **[See this comment for updated description of pull request](https://github.com/PennyLaneAI/pennylane/pull/2699#issuecomment-1163458859)**

**Context:**
There are very few simulators available today that offer qudit simulation. One notable example is [Google's Cirq](https://quantumai.google/cirq/qudits), which includes support for qudit operations. However, the simulator is far from complete, and a key drawback is the lack of operations available, which creates a need for users to define operations themselves. Popular libraries such as Qiskit and QuTiP also include some qudit manipulation capabilities, but with few use cases. Beyond these libraries, qudit simulation solutions are quite limited. Moreover, qudit simulation is a feature that is already in demand, as seen in this [issue](https://github.com/PennyLaneAI/pennylane/issues/2190) on the PennyLane repository.

This PR adds a `default.qutrit` device to PennyLane to showcase the ability to scale PennyLane to higher dimensions.

**Description of the Change:**
* Added new device `QutritDevice` as a parent device. Added `default.qutrit` for simulation of pure state qutrit devices. Users can currently make measurements with `qml.state()` and `qml.probs()` provided that an observable isn't passed into the measurements.
* Added operation `qml.QutritUnitary` for basic operation of `default.qutrit` device with user-specified unitaries

**Benefits:**
* Simulation of pure state qutrit circuits

**Possible Drawbacks:**
* Currently only supports `qml.QutritUnitary`, which requires user to provide unitary matrix as a parameter, so complex circuits can be tedious to implement.

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/2190

mudit2812 commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,800 @@
+
+        assert np.array_equal(res, np.array(expected))
+
+    test_binary_conversion_data = [
```
Yes, it should. The data itself is correct, however, the name of the list needs to be edited. I will add that change and commit once I resolve any remaining comments

mudit2812 commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,835 @@
+
+        if state is None:
+            raise qml.QuantumFunctionError("The state is not available in the current device")
+        # try:
+        #     state = self.state
+        # except:
+        #     raise qml.QuantumFunctionError("The state is not available in the current device")
```
It can. I added that for testing purposes and forgot to remove it. Thanks for pointing it out!

mudit2812 commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,835 @@
+import abc
```
This is a good point. Strictly considering code repetition, inheriting `QubitDevice` seems like a good idea. However, I'm a bit hesitant at the same time considering the mathematical differences that arise between qubits and qutrits. More conversation is definitely necessary, but I do agree that one of the possible final resolutions could be to dissolve both `QubitDevice` and `QutritDevice` in favor of a more general `QuditDevice`. That being said, a lot of work and discussion is necessary beforehand, as certain methods defined in `QubitDevice` and `QutritDevice`, such as `QubitDevice.states_to_binary()` will be harder to generalize and/or just be significantly slower once generalized.

glassnotes replied:
Maybe worth trying this if for no other reason than to shrink the PR and make it easier to review! It's at 3k lines, which is quite large. And inheriting in this way would also mean that many of the tests do not have to be rewritten. Would you be able to estimate how many lines of code could potentially be saved?

glassnotes replied:
The more I look at this, the more I think this is the right call for now. Maybe we should reopen the discussion in the ADR.

mudit2812 replied:
I would likely shrink `QutritDevice` by anywhere between 200-500 lines if I were to use the inheritance you're suggesting. I will update `QutritDevice` tomorrow and push the changes to the PR once they're tested

mudit2812 commented on doc/releases/changelog-dev.md:
```diff
@@ -4,40 +4,20 @@
 
 <h3>New features since last release</h3>
 
-* Added the new optimizer, `qml.SPSAOptimizer` that implements the simultaneous
```
I see. Thanks for letting me know!

mudit2812 commented on pennylane/devices/default_qutrit.py:
```diff
@@ -0,0 +1,317 @@
+        "QutritUnitary",
+    }
+
+    # TODO: Remove QutritUnitary from observables list
```
The code would break if the set of observables was empty as it was being interpreted to be an empty dict instead of a set.

mudit2812 commented on pennylane/devices/default_qutrit.py:
```diff
@@ -0,0 +1,317 @@
+        rotations = rotations or []
+
+        # apply the circuit operations
+        for i, operation in enumerate(operations):
```
The same syntax is in fact used in qubit device for functionality that I have not implemented yet. QubitDevice checks the index to make sure that state preparation operations aren't being applied **after** other operations have already been applied. However, I haven't added any operations for state preparations for qutrits yet.

mudit2812 commented on pennylane/ops/qutrit/__init__.py:
```diff
@@ -0,0 +1,15 @@
+
+# TODO: Remove QutritUnitary from obs list
+obs = {
+    "QutritUnitary",        # Added here to prevent errors when using device
```
It can't, as the list of observables must be limited to all observables defined inside the `/pennylane/ops/qutrit/` directory, and `Identity` is defined in `/pennylane/ops/`.

mudit2812 commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,90 @@
+
+    def adjoint(self):
+        U = self.matrix()
+        return QutritUnitary(qml.math.moveaxis(qml.math.conj(U), -2, -1), wires=self.wires)
```
Using `qml.math.moveaxis` with -2, -1 for the axes is the same as transposing the last two axes of a matrix, so in the case of a 2D array it behaves the same way as a matrix transpose. I don't entirely understand why this function is used instead of `qml.math.T`, but it behaves the same way regardless.

mudit2812 commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -0,0 +1,578 @@
+        with pytest.raises(qml.operation.PowUndefinedError):
+            op.pow(0.123)
+
+    # TODO: Check with Olivia about broadcasted QutritUnitary
```
That's what my first thought was as well. However, the test used for broadcasted operations in `/tests/ops/qubit/test_matrix_ops.py` doesn't use `qml.broadcast` and just passes a list of unitary matrices to `qml.QubitUnitary` (see [line 49-61](https://github.com/PennyLaneAI/pennylane/blob/5ce90652939fee5f1b0badd195ee20bb68fe0941/tests/ops/qubit/test_matrix_ops.py#L49-L61) in `/tests/ops/qubit/test_matrix_ops.py`), so I'm not entirely sure what "broadcasting" in that context means.

mudit2812 commented on pennylane/devices/default_qutrit.py:
```diff
@@ -0,0 +1,317 @@
+    def state(self):
+        return self._flatten(self._pre_rotated_state)
+
+    def density_matrix(self, wires):
```
The `density_matrix` method is available on the `default.qubit` device as well, and I've personally used it in the past, so I do see the benefit in keeping it.

mudit2812 commented on pennylane/devices/default_qutrit.py:
```diff
@@ -0,0 +1,317 @@
+        self._state = self._create_basis_state(0)
+        self._pre_rotated_state = self._state
+
+        # TODO: Add operations
```
I will do that now. Thanks for the suggestion!

mudit2812 commented on pennylane/devices/default_qutrit.py:
```diff
@@ -0,0 +1,317 @@
+
+        }
+
+    @functools.lru_cache()
```
Will do

mudit2812 commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,90 @@
+
+    Args:
+        U (array[complex]): square unitary matrix
+        wires(Sequence[int] or int): the wire(s) the operation acts on
```
I will add one soon.

mudit2812 commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,800 @@
+        assert circuit.device.shots == total_shots
+
+        # test gradient works
+        # TODO: Add after differentiability of qutrit circuits is implemented
```
I wasn't able to figure out exactly why `qml.jacobian` wasn't working. I will write a test that uses scalar parameters, but I don't see why a numpy array isn't working, as the `qml.jacobian` documentation says that they should work.

glassnotes replied:
What kind of error message were you getting? My initial thought is maybe it was the wrong NumPy, was it the one from PennyLane that was being imported when you tested?

mudit2812 replied:
Update: I've tried using a scalar input parameter to create a unitary matrix for differentiation, but the tests are still failing. My best guess is that the reason why is because `qml.QutritUnitary` doesn't currently support decomposition, so it can't be decomposed into parametrized operations that can be differentiated.

mudit2812 replied:
> What kind of error message were you getting? My initial thought is maybe it was the wrong NumPy, was it the one from PennyLane that was being imported when you tested?

Yes, I was using the PL version of numpy. I've included a screenshot of the errors below:
![image](https://user-images.githubusercontent.com/18223836/173707475-125d9cc3-e08b-4182-9341-95a2c2e26173.png)

mudit2812 commented:
Update: The `default.qutrit` device has been removed from this PR to decrease the size of the PR, and will be included in a separate PR in the coming weeks.

glassnotes commented on doc/releases/changelog-dev.md:
```diff
@@ -39,10 +39,23 @@
+  >>> U = np.array([[1, 1, 0], [1, -1, 0], [0, 0, np.sqrt(2)]]) / np.sqrt(2)
+  >>> with qml.tape.QuantumTape() as tape:
+  ...     qml.QutritUnitary(U, wires=0)
+  ...     return qml.probs(wires=0)
```
```suggestion
  ...     qml.probs(wires=0)
```
I don't think `return` is needed for tapes since this is a context and not a function. The example will also need to be updated to take into account the new PR structure.

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+        capabilities.update(model="qutrit")
+        return capabilities
+
+    # Added here for tests to pass (See tests.test_qutrit_device.ExtractStatistics.test_results_no_state())
```
Could you please provide the details of the error in a comment here?

mudit2812 replied:
Yes, I will update the comment.

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+        Returns:
+            array[int]: the sampled basis states
+        """
+        # A slower, but less memory intensive method
```
```suggestion
```
Comment not needed without the context of the faster method that is used in the qubit case.

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+        Returns:
+            array[int]: basis states in ternary representation
+        """
+        ternary_arr = []
```
I think the implementation here could be simplified by leveraging [`np.base_repr`](https://numpy.org/doc/stable/reference/generated/numpy.base_repr.html)? For example,
```
>>> np.base_repr(11, 3)
'102'
```

mudit2812 replied:
I agree that this would significantly simplify the implementation, but it would no longer be consistent with `QubitDevice.states_to_binary()`, which returns a nested list of 0s and 1s.

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+                    f"Unsupported return type specified for observable {obs.name}"
+                )
+
+            # TODO: Add support for MutualInfo and VnEntropy return types. Currently, qml.math is hard coded
```
What aspects of `qml.math` are hard-coded?

mudit2812 replied:
The helper functions in `qml.math` that calculate `VnEntropy` and `MutualInfo` have hardcoded 2's, which is why they can't be used to make those measurements for qutrits yet.

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+        sample_slice = Ellipsis if shot_range is None else slice(*shot_range)
+        samples = self._samples[sample_slice, device_wires]
+
+        # convert samples from a list of 0, 1 integers, to base 10 representation
```
```suggestion
        # convert samples from a list of 0, 1, 2 integers, to base 10 representation
```

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+            array[float]: list of the probabilities
+        """
+
+        wires = wires or self.wires
```
Aside from changing 2->3 in a few places, this function is identical to the qubit case as well, right?

mudit2812 replied:
Yep.

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+            indices = indices.reshape((bins, -1))
+            prob = np.zeros([3 ** len(device_wires), bins], dtype=np.float64)
+
+            # count the basis state occurrences, and construct the probability vector
```
```suggestion
```
Repeated comment?

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+        Returns:
+            array[float]: array of the resulting marginal probabilities.
+        """
+
```
In this function also, it looks like the only changes are a few 2 to 3?

mudit2812 replied:
Yep. Almost all of the functions that are overriden in `QutritDevice` only include changes to change a few 2's to 3's. One exception is `statistics()`, which has been overriden to not include the `VnEntropy` and `MutualInfo` return types for observables as they aren't currently supported on `QutritDevice`.

Speaking of those measurements, I just realized I should override them for `QutritDevice` as well such that their computation is explicitly not supported until later.

glassnotes commented on pennylane/_qutrit_device.py:
```diff
@@ -0,0 +1,390 @@
+        return samples.reshape((bin_size, -1))
+
+    # TODO: Implement function
+    def adjoint_jacobian(
```
It looks like this one also only needs a few 2->3 substitutions, but will need to wait until more operations are implemented.

mudit2812 replied:
Yes, I agree.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,103 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the discrete-variable quantum operations that
+accept a hermitian or an unitary matrix as a parameter.
```
```suggestion
This submodule contains the qutrit quantum operations that
accept a unitary matrix as a parameter.
```
(?)

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,103 @@
+"""
+# pylint:disable=abstract-method,arguments-differ,protected-access
+import warnings
+import numpy as np  # pylint: disable=unused-import
```
Why is NumPy being imported if it is not being used, or will it eventually come into play?

mudit2812 replied:
It was being used by ControlledQutritUnitary before I removed it from this PR. Thus, it will get used eventually, but for now I can remove it from the PR.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,103 @@
+    grad_method = None
+    """Gradient computation method."""
+
+    def __init__(self, *params, wires, do_queue=True):
```
It looks like there are a few updates that involve

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,103 @@
+
+        dim = 3 ** len(wires)
+
+        if qml.math.shape(U) != (dim, dim):
```
It looks like in the updated version of the `QubitUnitary`, there is also a dimension check for a valid batch size; it must be new, we should check with someone whether that needs to be implemented here as well.

mudit2812 replied:
Yes, I think that is new. It wasn't there when I initially implemented `QutritUnitary`, but it does make sense now as I recall seeing tests for `QubitUnitary` that used a list of unitary matrices as the input parameter for `QubitUnitary`. I'm a bit hesitant to add that feature just yet though because I haven't had a chance to look at the underlying implementation for batches in `QubitUnitary`, and I'm uncertain if it will just work out of the box.

Regardless, I do agree that getting input about whether it needs to be implemented would be helpful.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,103 @@
+    def adjoint(self):
+        U = self.matrix()
+        return QutritUnitary(qml.math.conj(qml.math.moveaxis(U, -2, -1)), wires=self.wires)
+
```
Noting here that eventually `_controlled` will be implemented along with `ControlledQubitUnitary`.

mudit2812 replied:
Yes, that is the plan. I'll add a comment to the code as well.

glassnotes reviewed (commented):
Comments on everything except the tests

glassnotes commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,1022 @@
+        with monkeypatch.context():
+            dev = mock_qutrit_device_extract_stats()
+            delattr(dev.__class__, "state")
+            del QubitDevice.state
```
This was the reason that `state` needed to be redefined in the `QutritDevice` class, right? Maybe provide some details of that here?

mudit2812 replied:
Yes, this was why.

glassnotes commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,1022 @@
+
+        expected = []
+
+        for s in samples:
```
This looks like essentially the same code that is being used within the `states_to_ternary` method; so if there is an error in the methodology, it wouldn't actually be caught. Recommend to either hardcode some examples in here, and/or to use a different built-in method in order to generate the expected values.

mudit2812 replied:
That is a good point. I'll change the test to use hardcoded examples.

glassnotes commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,1022 @@
+
+    @pytest.mark.parametrize("samples, ternary_states", test_ternary_conversion_data)
+    def test_correct_conversion(self, mock_qutrit_device, samples, ternary_states, tol):
+        """Tests that the states_to_binary method converts samples to binary correctly"""
```
```suggestion
        """Tests that the states_to_binary method converts samples to ternary correctly"""
```

glassnotes commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,1022 @@
+        res = dev.states_to_ternary(samples, wires)
+        assert np.allclose(res, ternary_states, atol=tol, rtol=0)
+
+
```
Note here that this is where the `TestExpval` and `TestVar` will be; there is also a `TestSample` in the qubit file, is that also coming afterwards or did I just miss it?

mudit2812 replied:
I added `TestSample` yesterday. While I can currently get 100% code coverage for `QutritDevice.sample()`, more tests will be added to it in the future to test sampling observables that have eigenvalues. No such observables for qutrits have been added to PL yet, so I will leave a comment in the code saying so.

glassnotes commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,1022 @@
+    """Tests for the execute method"""
+
+    def test_device_executions(self, mock_qutrit_device_extract_stats):
+        """Test the number of times a qubit device is executed over a QNode's
```
```suggestion
        """Test the number of times a qutrit device is executed over a QNode's
```

glassnotes commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,1022 @@
+            node_1(np.eye(3), np.eye(3), np.eye(9))
+        assert dev_1.num_executions == num_evals_1
+
+        # test a second instance of a default qubit device
```
```suggestion
        # test a second instance of a qutrit device
```

glassnotes commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,1022 @@
+            node_2(np.eye(3), np.eye(3))
+        assert dev_2.num_executions == num_evals_2
+
+        # test a new circuit on an existing instance of a qubit device
```
```suggestion
        # test a new circuit on an existing instance of a qutrit device
```

glassnotes commented on tests/test_qutrit_device.py:
```diff
@@ -0,0 +1,1022 @@
+        # res = qml.jacobian(circuit, argnum=[0])(pnp.eye(9, dtype=np.complex128))
+
+
+class TestUnimplemented:
```
Mention why these are necessary

mudit2812 replied:
They're necessary for complete coverage of `QutritDevice`. Some of the methods inherited from `QubitDevice` do not work out of the box for `QutritDevice`, and they have been overriden to raise a `NotImplementedError` when they are called, so I've added these tests to provide coverage for those methods. I will add a comment to the test file as well.

glassnotes reviewed (commented):
Have now gone briefly through the tests. This is pretty much ready to be marked as "ready for review" :tada: My only major recommendation for now is to update the overview of the PR, and add some summary notes to other devs as a comment on the PRs at the top of each relevant file, about: 
 - what is / is not the same as `QubitDevice`
 - what is currently missing that will come afterwards (and why it is currently missing, e.g., controlled functionality, expectation values, `DefaultQutrit`, etc.)

These details are throughout in the comments, it would just be helpful for review to see them all in one place.

mudit2812 commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -0,0 +1,376 @@
+        (unitary_group.rvs(9, random_state=10), 2),
+        (np.eye(3), 1),
+        (np.eye(9), 2),
+        # (np.tensordot([1j, -1, 1], U_thadamard_01, axes=0), 1)
```
This test uses a batch of unitaries rather than a single unitary matrix, and since that has not yet been implemented for `QutritUnitary`, it is not required. I will go back and remove that test for now from the PR and add it eventually once `QutritUnitary` supports batches of matrices.

mudit2812 commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -0,0 +1,376 @@
+]
+
+
+@pytest.mark.parametrize("mat, op", label_data)
```
Yes they are.

mudit2812 commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -0,0 +1,376 @@
+
+        mat = jnp.array([[1, 0, 0], [0, -1, 0], [0, 0, 1]])
+
+        self.check_interface(mat)
```
I'll do so.

glassnotes commented on doc/releases/changelog-dev.md:
````diff
@@ -39,10 +39,14 @@
   ```
-  
-* New PennyLane-inspired `sketch` and `sketch_dark` styles are now available for drawing circuit diagram graphics. 
+
+* New PennyLane-inspired `sketch` and `sketch_dark` styles are now available for drawing circuit diagram graphics.
````
```suggestion
  
* New PennyLane-inspired `sketch` and `sketch_dark` styles are now available for drawing circuit diagram graphics. 
```

glassnotes commented on doc/releases/changelog-dev.md:
```diff
@@ -58,4 +62,4 @@
 This release contains contributions from (in alphabetical order):
 
-Ankit Khandelwal, Ixchel Meza Chavez, Moritz Willmann
+Ankit Khandelwal, Ixchel Meza Chavez, Moritz Willmann, Mudit Pandey
```
```suggestion
Ankit Khandelwal, Ixchel Meza Chavez, Mudit Pandey, Moritz Willmann
```

glassnotes commented on doc/releases/changelog-dev.md:
```diff
@@ -39,10 +39,14 @@
+* New PennyLane-inspired `sketch` and `sketch_dark` styles are now available for drawing circuit diagram graphics.
   [(#2709)](https://github.com/PennyLaneAI/pennylane/pull/2709)
 
+* Added new device abstract base class `QutritDevice` for simulation of devices with wires of three dimensions. Users can currently make measurements with `qml.state()` and `qml.probs()`.
+* Added operation `qml.QutritUnitary` for applying user-specified unitaries to `QutritDevice`.
+
```
```suggestion
* Added new device abstract base class `QutritDevice` for simulation of devices with wires of three dimensions. Users can currently make measurements with `qml.state()` and `qml.probs()`.
  [(#2699)](https://github.com/PennyLaneAI/pennylane/pull/2699)  
* Added operation `qml.QutritUnitary` for applying user-specified unitaries to `QutritDevice`.
  [(#2699)](https://github.com/PennyLaneAI/pennylane/pull/2699)  
```

mudit2812 commented:
**Context:**
There are very few simulators available today that offer qudit simulation. One notable example is [Google's Cirq](https://quantumai.google/cirq/qudits), which includes support for qudit operations. However, the simulator is far from complete, and a key drawback is the lack of operations available, which creates a need for users to define operations themselves. Popular libraries such as Qiskit and QuTiP also include some qudit manipulation capabilities, but with few use cases. Beyond these libraries, qudit simulation solutions are quite limited. Moreover, qudit simulation is a feature that is already in demand, as seen in this [issue](https://github.com/PennyLaneAI/pennylane/issues/2190) on the PennyLane repository.

This PR adds a `QutritDevice` abstract qutrit device to PennyLane to showcase the ability to scale PennyLane to higher dimensions, along with a new operation `QutritUnitary`.

**Description of the Change:**
* Added new device `QutritDevice` as a parent device. Users can currently make measurements with `qml.state()` and `qml.probs()` provided that an observable isn't passed into the measurements.
* Added operation `qml.QutritUnitary` for basic operation of `QutritDevice` device with user-specified unitaries.

**Note:**
`QutritDevice` is an abstract base class, so all abstract methods will have to be manually set by a user before they can use `QutritDevice`.

`QutritDevice` inherits `QubitDevice`. This hierarchy makes sense due to two reasons:
* `QutritDevice` somewhat naturally scales into a higher dimension, so it makes sense for it to inherit the `QubitDevice` abstract class.
* `QutritDevice` and `QubitDevice` have very significant overlap between them, and a majority of the methods explicitly defined in `QutritDevice` are overrides with only have minor differences compared to `QubitDevice`. There are a few reasons for these overrides:
 * Difference in dimensions: Some methods in `QubitDevice` explicitly rely on the 2 dimensional nature of qubits, such as `estimate_probability()`, `marginal_prob()`, `sample()`, `generate_samples()`, and `generate_basis_states()`. These methods have been overridden to support 3-dimensions instead of 2.
 * Unsupported implementations: Some methods in `QubitDevice` are not supported by `QutritDevice` yet, such as `adjoint_jacobian()`, `density_matrix()`, `vn_entropy()`, `mutual_info()`, and they have been overridden  so that their `QubitDevice` implementations aren't accidentally used, which would cause a lot of unexpected behaviour. (Note: `density_matrix()`, `vn_entropy()`, and `mutual_info()` have been overridden because their helper functions in `qml.math` are hardcoded for two dimensions)
 * The `statistics()` method has been overridden because `QutritDevice` does not yet support some return types that `QubitDevice` supports, namely `VnEntropy`, `MutualInfo` and the density matrix.

**Benefits:**
* Abstract base class `QutritDevice` can be used as parent class for various qutrit PennyLane plugins.

**Possible Drawbacks:**
* Currently only supports `qml.QutritUnitary`, which requires user to provide unitary matrix as a parameter, so complex circuits can be tedious to implement.
* Users need to set the abstract methods of `QutritDevice` for it to be usable.

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/2190

antalszava commented:
Hi @mudit2812, thank you for the contribution! :slightly_smiling_face: :tada: 

The pull request seems to contain a large addition, could there be a way to split it up into parts so that there's only max. 800-1000 lines added? This would help us with doing code review.

mudit2812 commented:
> Hi @mudit2812, thank you for the contribution! 🙂 🎉
> 
> The pull request seems to contain a large addition, could there be a way to split it up into parts so that there's only max. 800-1000 lines added? This would help us with doing code review.

Hi @antalszava, thanks for your response. While it is _possible_ to make the PR smaller, a lot of functionality will definitely break. My thoughts are as follows:
* Removing `QutritUnitary` is probably not a great idea because at least one operation is needed for testing `QutritDevice`. While this can be solved by creating "test" operations, it will also limit what a user might currently be able to do with `QutritDevice`.
* Currently, all the methods defined in `QutritDevice` are overrides of methods in `QubitDevice`. A solution is to override those methods and raise a `NotImplementedError` for now. However, this will only get rid of ~200 lines of code in `pennylane/_qutrit_device.py`, but likely a lot more from `tests/test_qutrit_device.py`.
* ~70% of the additions are tests, which are very important for both ensuring correctness and reaching full coverage.

I will have a discussion with @glassnotes and update this comment with what the plan in moving forward for this PR.

mudit2812 commented:
**Context:**
There are very few simulators available today that offer qudit simulation. One notable example is [Google's Cirq](https://quantumai.google/cirq/qudits), which includes support for qudit operations. However, the simulator is far from complete, and a key drawback is the lack of operations available, which creates a need for users to define operations themselves. Popular libraries such as Qiskit and QuTiP also include some qudit manipulation capabilities, but with few use cases. Beyond these libraries, qudit simulation solutions are quite limited. Moreover, qudit simulation is a feature that is already in demand, as seen in this https://github.com/PennyLaneAI/pennylane/issues/2190 on the PennyLane repository.

This PR adds a new operation `QutritUnitary` to PennyLane for applying user-supplied unitaries to qutrit devices, which will be added in future pull requests.

**Description of the Change:**
* Added operation `qml.QutritUnitary` for basic operation of qutrit devices with user-specified unitaries.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/2190

glassnotes commented:
@mudit2812 thanks for paring this down, it looks much more manageable! Could you please update the CHANGELOG to reflect the new version? Other than that, this is ready for review from the Xanadu side :tada:

Qottmann commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -0,0 +1,378 @@
+        """Test QutritUnitary raised to a non-integer power raises an error."""
+        op = qml.QutritUnitary(U_thadamard_01, wires="a")
+
+        with pytest.raises(qml.operation.PowUndefinedError):
```
nice!

Qottmann reviewed (approved):
I dont have much to add to this, very nice and clean re-implementation of the qubit version of this 💯

mudit2812 commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,104 @@
+            U (tensor_like): unitary matrix
+
+        Returns:
+            tensor_like: canonical matrix
```
There isn't a qutrit device yet, so I can't add an example.

Qottmann replied:
Right, maybe just a comment "TODO" for later then :)

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,104 @@
+        U (array[complex]): square unitary matrix
+        wires(Sequence[int] or int): the wire(s) the operation acts on
+    """
+    num_wires = AnyWires
```
I'd leave it for consistency.

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -0,0 +1,378 @@
+            op.pow(0.123)
+
+    # TODO: Check with Olivia about broadcasted QutritUnitary
+    # def test_qutrit_unitary_noninteger_pow_broadcasted(self):
```
Tagging @dwierichs here since I see he is doing work with broadcasting :smile: 

David, what exactly does this broadcasting do, and is there a trick to getting the operations to work with it? I have found the `ndim_params` variable, but simply setting that (and doing a dimensionality check on the input) leads to dimensionality mismatches in the tests. Does some other part of the code need to be adjusted?

dwierichs replied:
Hi @glassnotes 
Without having looked through this PR carefully (very nice addition btw :) ) the following changes typically should be sufficient:
1. Add `ndim_params`
2. Make sure that `compute_decomposition`, if there is any, defers parameters in such a way that a leading dimension does not "get in the way", i.e. instead of `[qml.RX(x[0], wires=0), qml.RY(x[1], wires=1)]` do `[qml.RX(x[...,0], wires=0), qml.RY(x[...,1], wires=1)]` etc.
3. Make sure the operations that `compute_decomposition` returns also support broadcasting.
4. Write `compute_matrix` in such a way that it is compatible with receiving parameters that have an additional dimension compared to `ndim_params`. That is, `RX.compute_matrix` for example is able to take in a 1D tensor of parameters, and creates a 3-dimensional operation tensor, with the broadcasting dimension being the first. The straightforward way of doing this is to check `qml.math.ndim(param)` and do an if-else-loop.
5. Note that one also can just check `qml.math.ndim(param)`, compare with `ndim_params` and raise an Error if broadcasting attempted. Later we can add the functionality for broadcasting in a separate PR. (This error should optimally be raised both in `__init__` and in `compute_matrix`/`compute_decomposition`, I suppose).

mudit2812 replied:
Thanks for the insight! That's really helpful. I added the broadcasting related checks and tests for it as well.

glassnotes replied:
Thanks so much @dwierichs ! :tada:

glassnotes reviewed (approved):
Thanks for addressing all the suggestions @mudit2812 , this looks good to go from my end :100:

antalszava commented on pennylane/ops/qutrit/__init__.py:
```diff
@@ -0,0 +1,36 @@
+from ..identity import Identity
+
+ops = {
+    "Identity",
```
Will it not be an issue that `Identity` is implemented such that it's a qubit operation (e.g., its matrix representation)?

mudit2812 replied:
It will be in the future. However, currently there aren't any functions that explicitly rely on the matrix and/or eigenvalues of the operation/observable such as `expval` or `var`. This will need to be changed in the future, but for now, I will add a TODO, and I think we can leave it as is.

antalszava commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -0,0 +1,115 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
```
```suggestion
# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
```

antalszava reviewed (approved):
Looks good :tada: Awesome new type of addition to PennyLane! :slightly_smiling_face:
